### ISSUE TITLE
[Doc]: improve explanation of useAbility within useEffect, useMemo, ...

### ISSUE THREAD

**Issue by johannchopin-buyco** [stalniy/casl#756]
Hey @stalniy 👋 How are you doing?

I ran into an issue where the usage of `ability` within an effect and performance hook doesn't trigger an update:

```js
  const ability = useAbility(AbilityContext);

  React.useEffect(() => {
    console.log(ability);
  }, [ability]);
```

In this case doing an `ability.update` doesn't trigger the useEffect:

[![Edit ability-within-useEffect](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/s/ability-within-useeffect-zwflbb?fontsize=14&hidenavigation=1&theme=dark)

A real-life example would be with the usage of `useMemo`:

```js
  const posts = React.useMemo(() => getPosts(ability), [ability]); // get posts according to user abilities
  // ❌ ability.update won't update the list of posts
```

To trigger the change we can actually use the `ability.rules` that react can detect as "changed":
```js
  const posts = React.useMemo(() => getPosts(ability), [ability.rules]);
  // ✅ ability.update updates the list of posts
```

So If you agree I would love to add this usage to the react documentation.

**Comment by stalniy** [comment]
Yes, it triggers update only when new ability is provided via AbilityContext. So it’s still valid. And if you do not update ability and only rules, then you are correct. 

Happy to accept PR with relevant updates

**Comment by mdanka** [comment]
I agree that this is confusing because the documentation says that `useAbility` triggers a rerender when the rules are updated, which makes it sounds like that the same would happen to hooks depending on the ability. Thanks for making this issue, @johannchopin-buyco !
